plugins/woocommerce-blocks/assets/js/atomic/utils/register-block-single-product-template.ts

```diff
@@ -1,140 +1,319 @@
-					previousTemplateId?.includes( 'single-product' ) )
-			) {
-				if ( isVariationBlock && variationName ) {
-					unregisterBlockVariation( blockName, variationName );
-				} else {
-					unregisterBlockType( blockName );
+			// Only run this once for site editor
+			siteEditorUnsubscribe();
+
+			// Set up the template change listener
+			subscribe( () => {
+				const previousTemplateId = this.currentTemplateId;
+				this.currentTemplateId = this.parseTemplateId(
+					editSiteStore?.getEditedPostId<
+						string | number | undefined
+					>()
+				);
+
+				if ( previousTemplateId !== this.currentTemplateId ) {
+					this.handleTemplateChange( previousTemplateId );
 				}
-				isRegistered = false;
+			}, 'core/edit-site' );
+
+			// Initial registration of blocks for site editor
+			this.blocks.forEach( ( config ) => {
+				this.registerBlock( config );
+			} );
+		}, 'core/edit-site' );
+
+		// Post Editor Subscription
+		const postEditorUnsubscribe = subscribe( () => {
+			const editPostStore = select( 'core/edit-post' );
+			if ( ! editPostStore ) {
+				return;
 			}
 
-			if ( ! isRegistered ) {
-				if ( isVariationBlock ) {
-					// @ts-expect-error: `registerBlockType` is not typed in WordPress core
-					registerBlockVariation( blockName, blockSettings );
-				} else {
-					const ancestor = isEmpty( blockSettings?.ancestor )
-						? [ 'woocommerce/single-product' ]
-						: blockSettings?.ancestor;
-					// @ts-expect-error: `registerBlockType` is not typed in WordPress core
-					registerBlockType( blockMetadata, {
-						...blockSettings,
-						ancestor: ! currentTemplateId?.includes(
-							'single-product'
-						)
-							? ancestor
-							: undefined,
-					} );
+			// Only run this once for post editor
+			postEditorUnsubscribe();
+
+			// Register blocks that are available in post editor
+			this.blocks.forEach( ( config ) => {
+				if ( config.isAvailableOnPostEditor ) {
+					const key = config.variationName || config.blockName;
+					if ( ! this.registeredBlocks.has( key ) ) {
+						this.registerBlock( config );
+						this.registeredBlocks.add( key );
+					}
 				}
+			} );
+		}, 'core/edit-post' );
+
+		this.initialized = true;
+	}
+
+	/**
+	 * Handles block registration/unregistration when template context changes.
+	 * Specifically manages transitions to and from single product templates.
+	 *
+	 * @param {string | undefined} previousTemplateId - The previous template ID
+	 */
+	private handleTemplateChange(
+		previousTemplateId: string | undefined
+	): void {
+		const isTransitioningToOrFromSingleProduct =
+			this.currentTemplateId?.includes( 'single-product' ) ||
+			previousTemplateId?.includes( 'single-product' );
+
+		if ( ! isTransitioningToOrFromSingleProduct ) {
+			return;
+		}
+
+		this.blocks.forEach( ( config ) => {
+			if ( ! this.isRegistering ) {
+				this.unregisterBlock( config );
+				this.registerBlock( config );
 			}
-		}, 'core/edit-site' );
+		} );
```

Comment:
We are coupling the `BlockRegistrationManager` to the single-product template. Since this is our only current use case, I think that it is fine. Let's note this in the documentation.

Reply:
Good idea, done in https://github.com/woocommerce/woocommerce/pull/53788/commits/b54cd2d19f3d0e79bdeea1d4c902d53bf904bd71